src/Tribe/Rewrite.php

```diff
@@ -0,0 +1,311 @@
+<?php
+/**
+ * Rewrite Configuration Class
+ */
+
+// Don't load directly
+if ( ! defined( 'ABSPATH' ) ) {
+	die( '-1' );
+}
+
+if ( ! class_exists( 'Tribe__Events__Rewrite' ) ) {
+
+	/**
+	 * Permalinks magic Happens over here!
+	 */
+	class Tribe__Events__Rewrite {
+
+		/**
+		 * Static singleton variable
+		 * @var [type]
+		 */
+		public static $instance;
+
+		/**
+		 * Static Singleton Factory Method
+		 *
+		 * @return Tribe__Events__Rewrite
+		 */
+		public static function instance( $wp_rewrite = null ) {
+			if ( ! isset( self::$instance ) ) {
+				self::$instance = new self( $wp_rewrite );
+			}
+
+			return self::$instance;
+		}
+
+		/**
+		 * WP_Rewrite Instance
+		 * @var WP_Rewrite
+		 */
+		public $rewrite;
+
+		/**
+		 * Rewrite rules Holder
+		 * @var array
+		 */
+		public $rules = array();
+
+		/**
+		 * Base slugs for rewrite urls
+		 * @var array
+		 */
+		public $bases = array();
+
+		private function __construct() {}
+
+		public function hooks() {
+			add_filter( 'generate_rewrite_rules', array( $this, '_filter_generate' ) );
+		}
+
+		public function _filter_generate( WP_Rewrite $wp_rewrite ) {
```

Comment:
Yeah, let's not prefix functions with `_` and instead use `protected` and `private` where necessary.